src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -528,26 +528,31 @@ std::unique_ptr<Uptane::Target> SotaUptaneClient::findTargetInDelegationTree(con
   return findTargetHelper(*toplevel_targets, target, 0, false);
 }
 
-result::Download SotaUptaneClient::downloadImages(const std::vector<Uptane::Target> &targets,
+result::Download SotaUptaneClient::downloadImages(std::vector<Uptane::Target> targets,
                                                   const api::FlowControlToken *token) {
   // Uptane step 4 - download all the images and verify them against the metadata (for OSTree - pull without
   // deploying)
   std::lock_guard<std::mutex> guard(download_mutex);
   result::Download result;
   std::vector<Uptane::Target> downloaded_targets;
-  for (auto it = targets.cbegin(); it != targets.cend(); ++it) {
-    auto images_target = findTargetInDelegationTree(*it);
+  for (auto &it : targets) {
+    auto images_target = findTargetInDelegationTree(it);
     if (images_target == nullptr) {
       // TODO: Could also be a missing target or delegation expiration.
-      last_exception = Uptane::TargetMismatch(it->filename());
-      LOG_ERROR << "No matching target in images targets metadata for " << *it;
+      last_exception = Uptane::TargetMismatch(it.filename());
+      LOG_ERROR << "No matching target in images targets metadata for " << it;
       result = result::Download(downloaded_targets, result::DownloadStatus::kError, "Target hash mismatch.");
       sendEvent<event::AllDownloadsComplete>(result);
       return result;
     }
+    // If the URL from the Director is unset, but the URL from the Images repo
+    // is set, use that.
+    if (it.uri().empty() && !images_target->uri().empty()) {
+      it.setUri(images_target->uri());
+    }
   }
-  for (auto it = targets.cbegin(); it != targets.cend(); ++it) {
-    auto res = downloadImage(*it, token);
+  for (const auto &it : targets) {
+    auto res = downloadImage(it, token);
     if (res.first) {
       downloaded_targets.push_back(res.second);
     }
```

Comment:
Why do we need to continue downloading if a current target image download fails, don't we consider a singe multi-target update as an atomic operation  ?

Reply:
Not necessarily. That logic is still being worked out as far as I understand. That should be up to the user to decide what to do. For now, we attempt to download everything regardless of failures.

Reply:
Maybe, makes sense to make it configurable.

Reply:
Agreed. That requires some integration with the backend, though, and we haven't sorted out the details yet.